RESOURCE_PROVIDER_PLUGINS.md

```diff
@@ -0,0 +1,299 @@
+# Resource Provider plugins
+
+## Overview
+
+This document should be read in conjunction with [SERVERPLUGINS.md](./SERVERPLUGINS.md) as it contains additional information regarding the development of plugins that facilitate the storage and retrieval of resource data (e.g. routes, waypoints, notes, regions and charts).
+
+Resource storage and retrieval is de-coupled from core server function to provide the flexibility to implement the appropriate resource storage solution for specific Signal K implementations.
+
+The Signal K Node server will pass requests made to the following paths  to registered resource providers:
+- `/signalk/v1/api/resources`
+- `/signalk/v1/api/resources/routes`
+- `/signalk/v1/api/resources/waypoints`
+- `/signalk/v1/api/resources/notes`
+- `/signalk/v1/api/resources/regions`
+- `/signalk/v1/api/resources/charts`
+
+Resource providers will receive request data via a `ResourceProvider` interface which they implement. It is the responsibility of the resource provider to persist resource data in storage (PUT, POST), retrieve the requested resources (GET) and remove resource entries (DELETE).
+
+Resource data passed to the resource provider plugin has been validated by the server and can be considered ready for storage.
+
+
+## Resource Providers
+
+A `resource provider plugin` is responsible for the storage and retrieval of resource data.
+
+It should implement the necessary functions to:
+- Persist each resource with its associated id
+- Retrieve an individual resource with the supplied id
+- Retrieve a list of resources that match the supplied qery criteria.
+
+Data is passed to and from the plugin via the methods defined in the  __resourceProvider__ interface which the plugin must implement.
+
+_Definition: `resourceProvider` interface._
+```javascript
+resourceProvider: {
+  types: [],
+  methods: {
+    listResources: (type:string, query: {[key:string]:any})=> Promise<any>
+    getResource: (type:string, id:string)=> Promise<any>
+    setResource: (type:string, id:string, value:{[key:string]:any})=> Promise<any>
+    deleteResource: (type:string, id:string)=> Promise<any>
+  }
+}
+```
+
+This interface is used by the server to direct requests to the plugin. 
+
+It contains the following attributes:
+- `types`: An array containing the names of the resource types the plugin is a provider for.  Names of the resource types are: `routes, waypoints, notes, regions, charts`.
+
+- `methods`: The methods to which the server dispatches requests. The plugin will implement these methods to perform the necessary save or retrieval operation. Each method returns a promise containing either resource data or `null` if an error is encountered.
```

Comment:
Imho the Promise should throw, not return null.

Reply:
I will update to resolved or rejected Promise

Reply:
Updated to return:
- Resolved Promise on success
- Rejected Promise on incompletion.